mullvad-daemon/src/tunnel.rs

```diff
@@ -139,126 +144,143 @@ impl ParametersGenerator {
-            Err(_error) => Err(Error::NoRelayAvailable),
-        }
-    }
-
-    #[cfg_attr(target_os = "android", allow(unused_variables))]
-    async fn create_tunnel_parameters(
-        &mut self,
-        relay: &Relay,
-        entry_relay: &Option<Relay>,
-        endpoint: MullvadEndpoint,
-        bridge: Option<SelectedBridge>,
-        obfuscator: Option<SelectedObfuscator>,
-    ) -> Result<TunnelParameters, Error> {
         let data = self.device().await?;
-        match endpoint {
+        let selected_relay = self
+            .relay_selector
+            .get_relay(retry_attempt as usize)
+            .map_err(|err| match err {
+                mullvad_relay_selector::Error::NoBridge => Error::NoBridgeAvailable,
+                _ => Error::NoRelayAvailable,
+            })?;
+
+        match selected_relay {
             #[cfg(not(target_os = "android"))]
-            MullvadEndpoint::OpenVpn(endpoint) => {
-                let (bridge_settings, bridge_relay) = match bridge {
-                    Some(SelectedBridge::Normal(bridge)) => {
-                        (Some(bridge.settings), Some(bridge.relay))
-                    }
-                    Some(SelectedBridge::Custom(settings)) => (Some(settings), None),
-                    None => (None, None),
+            GetRelay::OpenVpn {
+                endpoint,
+                exit,
+                bridge,
+            } => {
+                let bridge_settings = match bridge {
+                    Some(SelectedBridge::Normal { ref settings, .. }) => Some(settings.clone()),
+                    Some(SelectedBridge::Custom(ref settings)) => Some(settings.clone()),
+                    None => None,
+                };
+                let bridge_relay = match bridge {
+                    Some(SelectedBridge::Normal { relay, .. }) => Some(relay),
+                    _ => None,
                 };
 
+                // TODO(markus): Can this be done 'generically'?
                 self.last_generated_relays = Some(LastSelectedRelays::OpenVpn {
-                    relay: relay.clone(),
+                    relay: exit.clone(),
                     bridge: bridge_relay,
                 });
 
-                Ok(openvpn::TunnelParameters {
-                    config: openvpn::ConnectionConfig::new(
-                        endpoint,
-                        data.account_token,
-                        "-".to_string(),
-                    ),
-                    options: self.tunnel_options.openvpn.clone(),
-                    generic_options: self.tunnel_options.generic.clone(),
-                    proxy: bridge_settings,
-                    #[cfg(target_os = "linux")]
-                    fwmark: mullvad_types::TUNNEL_FWMARK,
-                }
-                .into())
+                // TODO(markus): Remodel `GetRelay` to not require this unwrap.
+                Ok(self.create_openvpn_tunnel_parameters(
+                    *endpoint.unwrap_openvpn(),
```

Comment:
Just a thought. Why does `GetRelay::OpenVpn.endpoint` have type `MullvadEndpoint` and not `Endpoint` since it can't be the `Wireguard(MullvadWireguardEndpoint)` variant?

Reply:
It doesn't! I had totally forgotten about this comment :sweat_smile: Swapping `MullvadEndpoint` out for `Endpoint` made a lot of code more straightforward. Thanks for the suggestion!